File: app/src/main/scala/algorithm/lc/bt/BinaryTreePostOrderTraversal.java
package algorithm.lc.bt;

import algorithm.lc.bt.common.TreeNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * 后续遍历
 */
public class BinaryTreePostOrderTraversal {
    public static List<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> res = new ArrayList<>();
        Stack<TreeNode> stacks = new Stack<>();

        stacks.push(root);

        while (!stacks.isEmpty()) {
            TreeNode pop = stacks.pop();

            res.add(pop.val);
            if (pop.left != null) {
                stacks.push(pop.left);
            }

            if (pop.right != null) {
                stacks.push(pop.right);
            }
        }

        return res;
    }
}

File: app/src/main/scala/algorithm/lc/BinaryTreePostOrderTraversal.scala
package algorithm.lc

import scala.collection.mutable

object BinaryTreePostOrderTraversal {
  /**
   * 后序遍历
   */
  def preorderTraversal(root: TreeNode): List[Int] = {
    var nodes = new mutable.Stack[TreeNode]()
    var res = List[Int]()
    if (root == null) {
      return res
    }
    nodes.push(root)
    while (nodes.nonEmpty) {
      val node = nodes.pop()
      res = res :+ node.value
      if (node.right != null) nodes.push(node.right)
      if (node.left != null) nodes.push(node.left)
    }
    res
  }

  class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
    var value: Int = _value
    var left: TreeNode = _left
    var right: TreeNode = _right
  }

}
